Further improvement of the SSG test suite

matejak opened the pull request:
This PR builds on top of #3179:

* With recent-enough `oscap-ssh`, execute SSH-based operations completely disregarding `known_hosts` - see also https://github.com/OpenSCAP/openscap/pull/1177.
* Refactored the state handling: There is a new class `SavedState` that maps to stored persistent state (docker image, VM snapshot). The class has a `map_on_top` method, that fits the purpose of starting from the defined state, performing state alternations and scan, and reverting to the original state to run another check.
* Introduced offline scanning for libvirt and docker backends using `oscap-vm` and `oscap-docker` respectively. `oscap-vm` doesn't work, which is a known issue, and `oscap-docker` may require the test suite to be executed by the root user.
* Introduced the `common` module with shared utility functions..
* Dropped usage of a `scenario` saved state that was not needed - we just need the state with all tests scripts uploaded.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/OpenSCAP/scap-security-guide/inspections/80417351-5098-4490-83f4-e02c61725403) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/OpenSCAP/scap-security-guide/inspections/80417351-5098-4490-83f4-e02c61725403) completed: **1 new issues, 28 updated code elements**

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/OpenSCAP/scap-security-guide/inspections/f825b2c9-e045-46d4-895f-e07f62fad9ed) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/OpenSCAP/scap-security-guide/inspections/f825b2c9-e045-46d4-895f-e07f62fad9ed) completed: **42 updated code elements**

dahaic commented:
Jenkins test this please.

jan-cerny commented on tests/README.md:
```diff
@@ -192,7 +195,7 @@ Following services need to be supported:
 - You may want to include another packages, as base images tend to be bare-bone and tests may require more packages to be present.
 Also, as containers may get any IP address, a conflict may appear in your local client's `known_hosts` file.
-As `oscap-ssh` doesn't let you to switch known hosts checks off at the client-side, it may be a good idea to disable known hosts checks for all hosts if you are testing on a VM or under a separate user.
+You might have a version of `oscap-ssh` that doesn't support ssh connetion customization at the client-side, so it may be a good idea to disable known hosts checks for all hosts if you are testing on a VM or under a separate user.
```
connetion

jan-cerny commented on tests/ssg_test_suite/test_env.py:
```diff
@@ -2,81 +2,188 @@
+    def _delete_saved_state(self, snapshot):
         self.snapshot_stack.revert()
 
+    def _oscap_docker_base_arguments(self):
```
I think this function has a confusing name because it is using `oscap-vm` inside, not oscap-docker.

jan-cerny commented on tests/ssg_test_suite/test_env.py:
```diff
@@ -2,81 +2,188 @@
+        command_base.extend(
+            ['oscap-vm', "domain", self.domain_name])
+        command_base.extend(
+            ['xccdf', 'eval'])
```
I would do it at once.

jan-cerny commented on tests/ssg_test_suite/test_env.py:
```diff
@@ -2,81 +2,188 @@
+        command_base.extend(
+            ['oscap-ssh', full_hostname, '22'])
+        command_base.extend(
+            ['xccdf', 'eval'])
```
I would do everything at once.

jan-cerny commented on tests/ssg_test_suite/oscap.py:
```diff
@@ -299,10 +273,18 @@ def _generate_report_file(self):
 
-    def prepare_oscap_ssh_arguments(self):
-        full_hostname = 'root@{}'.format(self.domain_ip)
+    def prepare_oscap_docker_arguments(self):
```
This function is not used anywhere.

jan-cerny commented on tests/ssg_test_suite/oscap.py:
```diff
@@ -299,10 +273,18 @@ def _generate_report_file(self):
+    def prepare_oscap_docker_arguments(self):
         self.command_base.extend(
-            ['oscap-ssh', full_hostname, '22', 'xccdf', 'eval'])
+            ['oscap-docker', "xxx", 'xccdf', 'eval'])
```
xxx?

jan-cerny commented on tests/ssg_test_suite/rule.py:
```diff
@@ -55,20 +57,19 @@ def get_viable_profiles(selected_profiles, datastream, benchmark):
-def _run_with_stdout_logging(command, log_file):
-    log_file.write(" ".join(command) + "\n")
+def _run_with_stdout_logging(command, args, log_file):
+    log_file.write(" ".join((command,) + args) + "\n")
```
This function is always used with first parameter as a string:
```
if _run_with_stdout_logging("ssh", args, log_file) 
if _run_with_stdout_logging("scp", args, log_file) 
````
I think it would be more readable this way:
```
log_file.write("{0} {1}\n".format(command, " ".join(args))
```
That way we would avoid commas.

jan-cerny commented on tests/ssg_test_suite/rule.py:
```diff
@@ -110,14 +113,12 @@ def _apply_script(rule_dir, domain_ip, script):
+        args = common.IGNORE_KNOWN_HOSTS_OPTIONS + (machine, command)
+
+        rc = _run_with_stdout_logging("ssh", args, log_file)
+        if rc != 0:
```
This way reminds me programming in C. However, in Python we use exception to handle problems.

jan-cerny commented on tests/test_suite.py:
```diff
@@ -56,6 +56,15 @@ def parse_args():
+        choices=("online", "offline"),
+        help="What type of check to use: "
+        "Online check done by running oscap inside the concerned system, or "
+        "offline check that examines the filesystem from the host?")
```
My opinion is that this could be reworded in a way there is no question.

jan-cerny commented:
In the new offline mode it does weird fails:
```[jcerny@thinkpad tests{ssgts-turbo2}]$ python3 test_suite.py rule --docker ssg_test_suite --mode offline --datastream ../build/ssg-centos7-ds.xml --xccdf-id scap_org.open-scap_cref_ssg-rhel7-xccdf-1.2.xml rule_sshd_disable_kerb_auth
Setting console output to log level INFO
INFO - The base image option has been specified, choosing Docker-based test environment.
INFO - Logging into /home/jcerny/scap-security-guide/tests/logs/rule-custom-2018-08-28-0950/test_suite.log
INFO - xccdf_org.ssgproject.content_rule_sshd_disable_kerb_auth
ERROR - Failed to cleanup file '/home/jcerny/scap-security-guide/tests/logs/rule-custom-2018-08-28-0950/xccdf_org.ssgproject.content_rule_sshd_disable_kerb_auth-wrong_value.fail.sh-initial.html'
ERROR - Script wrong_value.fail.sh using profile xccdf_org.ssgproject.content_profile_ospp found issue:
ERROR - Scan has exited with return code 0, instead of expected 2 during stage initial
ERROR - The initial scan failed for rule 'xccdf_org.ssgproject.content_rule_sshd_disable_kerb_auth'.
ERROR - Failed to cleanup file '/home/jcerny/scap-security-guide/tests/logs/rule-custom-2018-08-28-0950/xccdf_org.ssgproject.content_rule_sshd_disable_kerb_auth-correct_value.pass.sh-initial.html'
INFO - Script correct_value.pass.sh using profile xccdf_org.ssgproject.content_profile_ospp OK
ERROR - Failed to cleanup file '/home/jcerny/scap-security-guide/tests/logs/rule-custom-2018-08-28-0950/xccdf_org.ssgproject.content_rule_sshd_disable_kerb_auth-comment.pass.sh-initial.html'
INFO - Script comment.pass.sh using profile xccdf_org.ssgproject.content_profile_ospp OK
ERROR - Failed to cleanup file '/home/jcerny/scap-security-guide/tests/logs/rule-custom-2018-08-28-0950/xccdf_org.ssgproject.content_rule_sshd_disable_kerb_auth-line_not_there.pass.sh-initial.html'
INFO - Script line_not_there.pass.sh using profile xccdf_org.ssgproject.content_profile_ospp OK
[jcerny@thinkpad tests{ssgts-turbo2}]$ 
```

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/OpenSCAP/scap-security-guide/inspections/d0f1c252-4af6-41c1-b746-7c0bf3eaab6f) was created.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/OpenSCAP/scap-security-guide/inspections/a0906d88-a99f-4fad-89f5-347cfbbbbbad) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/OpenSCAP/scap-security-guide/inspections/d0f1c252-4af6-41c1-b746-7c0bf3eaab6f) completed: **41 updated code elements**

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/OpenSCAP/scap-security-guide/inspections/a0906d88-a99f-4fad-89f5-347cfbbbbbad) completed: **41 updated code elements**

matejak commented:
@jan-cerny Thank you for your review! I think that I have addressed your comments, as I agree with them.
I have added more information to the `README` - you have to run the offline `docker`-based  checks with `sudo` (and you therefore need an image that accepts `root`'s preferably passwordless SSH key.
Unfortunately, due to an anti-feature of `oscap-docker`, false positives are going to be reported by the scan, unless `oscap-docker` from https://github.com/OpenSCAP/openscap/pull/1187 is used.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/OpenSCAP/scap-security-guide/inspections/b4a00ede-8078-4e01-a271-67ee84479399) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/OpenSCAP/scap-security-guide/inspections/b4a00ede-8078-4e01-a271-67ee84479399) completed: **41 updated code elements**

matejak commented:
jenkins test this please

matejak commented:
jenkins test this please

matejak commented:
jenkins test this please
